upx-recovery-tool.py

```diff
@@ -0,0 +1,289 @@
+            br"^\x18\xd0\x4d\xe2.{2}\x00\xeb\x00\xc0\xdd\xe5\x0e\x00\x5c\xe3.\x02\x00\x1a\x0c\x48\x2d\xe9\x00\xb0\xd0\xe5\x06\xcc\xa0\xe3",
+            br"^\x18\xd0\x4d\xe2.{2}\x00\xeb\x00\x10\x81\xe0\x3e\x40\x2d\xe9\x00\x50\xe0\xe3\x02\x41\xa0\xe3.{2}\x00\xea\x1a\x00\xbd\xe8",
+        ],
+        "PowerPC or cisco 4500": [
+            # ucl
+            br"^\x48\x00.{2}\x7c\x00\x29\xec\x7d\xa8\x02\xa6\x28\x07\x00\x02\x40\x82\x00\xe4\x90\xa6\x00\x00",
+            # lzma
+            br"^\x48\x00.{2}\x28\x07\x00\x0e\x40\x82\x0a\x4c\x94\x21\xff\xe8\x7c\x08\x02\xa6\x7c\xc9\x33\x78\x81\x06\x00\x00\x7c\xa7\x2b\x78",
+        ],
+    }
+
+    def __init__(self, in_file, out_file):
+        """ Initialization method. Receives the file to be read and the output path to write the fixed UPX """
+        
+        self.in_file = in_file
+        self.out_file = out_file
+
+        # Check file type is ELF and arch is supported
+        self.check_file_type()
+
+        # File is ELF, so it can be parsed
+        self.in_fd = open(self.in_file, "rb")
+        self.elf = ELFFile(self.in_fd)
+
+        # Check if is UPX
+        if not self.is_upx():
+            raise NonUpxError
+
+        # Get UPX version. p_info fix doesn't work with UPX 4
+        self.detect_version()
+
+        # All checks passed
+        shutil.copy(in_file, out_file)
+
+        self.out_fd = open(out_file, "r+b")
+        self.buff = mmap.mmap(self.out_fd.fileno(), 0)
+
+        self.load_structs()
+
+    def check_file_type(self):
+        """ Method to check if the class will be able to analyze this type of file """
+
+        if not os.path.isfile(self.in_file):
+            raise UnsupportedFileError("Is not a file")
+
+        # Check magic filetype
+        magic_str = magic.from_file(self.in_file)
+        
+        if magic_str.startswith("ELF "):
+            self.exe_type = "ELF"
+            magic_arch = magic_str.split(',')[1].strip()
+
+            if magic_arch in self.upx_sigs.keys():
+                self.arch = magic_arch
+                return
+
+        raise UnsupportedFileError(f"Unsupported file type '{magic_str}'")
+
+    def is_upx(self):
+        """ Method that looks for ASM signatures to identify an UPX executable """
+
+        # Instead of looking for the pattern in all the executable bytes, look for the
+        # UPX sigs in the EP. The code is not so beauty, but it'll be more efficient.
+        ep_bytes = self.get_ep_bytes(50)
+
+        for sig in self.upx_sigs[self.arch]:
+            if re.match(sig, ep_bytes, re.DOTALL):
+                return True
+
+        return False
+
+    def detect_version(self):
+        """ Method to know the UPX version used to pack the executable """
+        # TODO: The current detection method is very naive and can be easily tricked
+
+        # Load file in memory to look for string with version
+        self.in_fd.seek(0)
+        in_buff = mmap.mmap(self.in_fd.fileno(), 0, access=mmap.ACCESS_READ)
```

Comment:
just curious, is there any advantage of this approach compared to the classic read() method?